pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -253,8 +304,202 @@ void Stencil::CollectStrideOneIndices() {
+      }
+      for (unsigned i = 0; i < kNumIndex; ++i) {
+        bestIdxs[i] = innerIdxs[i];
+        bestTiles[i] = tiles[i];
+      }
+    }
+    return;
+  }
+
+  unsigned range = idxRange(innerIdxs[idx]);
+  for (unsigned i = range; i > 0; --i) {
+    if (onlyEven[idx] && (range % i != 0)) {
+      continue;
+    }
+    tiles[idx] = i;
+    SearchTiles(idx + 1);
+  }
+}
+
+int64_t Stencil::idxRange(mlir::BlockArgument idx) {
+  assert(idx.getArgNumber() < opConstRanges.size());
+  return opConstRanges[idx.getArgNumber()];
+}
+
+double Stencil::Evaluate() {
+  unsigned tot_inner_loop = tiles[0] * tiles[1] * tiles[2];
+  double throughput;
+  unsigned startup_cost;
+  std::tie(throughput, startup_cost) = Throughput(tiles[0], tiles[1], tiles[2]);
+  if (throughput == 0) {
+    return std::numeric_limits<double>::max();
+  }
+  double inner_time = tot_inner_loop / throughput;
+  IVLOG(3,
+        "Inner: loop = " << tot_inner_loop << " inner_time = " << inner_time);
+  for (unsigned i = 0; i < kNumIndex; ++i) {
+    IVLOG(3, innerIdxs[i] << ": " << tiles[i]);
+  }
+
+  llvm::DenseMap<mlir::BlockArgument, unsigned> middle_idxs;
+  for (auto idx : accIdxs) {
+    middle_idxs.try_emplace(idx, idxRange(idx));
+  }
+  for (unsigned i = 0; i < kNumIndex; ++i) {
+    auto it = middle_idxs.find(innerIdxs[i]);
+    if (it != middle_idxs.end()) {
+      it->second = (it->second - 1) / tiles[i] + 1;
+    }
+  }
+  unsigned tot_middle_loop = 1;
+  for (auto &kvp : middle_idxs) {
+    tot_middle_loop *= kvp.second;
+  }
+
+  IVLOG(3, "Middle: loop = " << tot_middle_loop);
+
+  for (auto &kvp : middle_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, kvp.first << ": " << kvp.second);
+    }
+  }
+
+  llvm::DenseMap<mlir::BlockArgument, unsigned> outer_idxs;
+  for (auto idx : outIdxs) {
+    outer_idxs.try_emplace(idx, idxRange(idx));
+  }
+  for (unsigned i = 0; i < kNumIndex; ++i) {
+    auto it = outer_idxs.find(innerIdxs[i]);
+    if (it != outer_idxs.end()) {
+      it->second = (it->second - 1) / tiles[i] + 1;
+    }
+  }
+  unsigned tot_outer_loop = 1;
+  for (auto &kvp : outer_idxs) {
+    tot_outer_loop *= kvp.second;
+  }
+
+  IVLOG(3, "Outer: loop = " << tot_outer_loop);
+
+  for (auto &kvp : outer_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, kvp.first << ": " << kvp.second);
+    }
+  }
+
+  unsigned outer_batches =
+      (tot_outer_loop - 1) / std::thread::hardware_concurrency() + 1;
+  double perf = outer_batches * tot_middle_loop * (startup_cost + inner_time);
+
+  IVLOG(3, "Performance = " << perf);
+  return perf;
+}
+
+std::pair<double, unsigned> Stencil::Throughput(unsigned m, unsigned n,
+                                                unsigned k) {
+  // TODO: Hook up heatmap code.
+  return std::make_pair(3.0, 32);
+}
+
+void Stencil::SetNumberThreads() {
+  if (numThreads == 0) {
+    numThreads = std::thread::hardware_concurrency() + 1;
+  }
+}
+
 void Stencil::DoStenciling() {
   // Initialization
+  SetNumberThreads();
```

Comment:
We should construct the `Stencil` object with whatever options it needs. And we should have a `Stencil` object per pass.

Basically, the `Stencil` object should be setup by construction.

Reply:
So, the design we talked over with Jeremy is that the Stencil object correspond to a AffineParallelOp. So, there are many Stencil-ers per pass - one for each AffineParralelOp. And each Stencil stencils a single AffineParallelOp.

Reply:
This seems orthogonal to the request to construct a `Stencil` explicitly with the numberOfThreads set instead of setting it after the `Stencil` object has been constructed.

Reply:
The threads option is initialized from outside now.